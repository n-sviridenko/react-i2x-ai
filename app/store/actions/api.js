import { createAction } from 'redux-actions';

export const REQUEST_FAIL = 'app/api/REQUEST_FAIL';

export const requestFail = createAction(REQUEST_FAIL);
