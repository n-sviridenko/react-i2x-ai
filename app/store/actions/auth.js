import { createAction } from 'redux-actions';
import { Map } from 'immutable';

export const LOGIN_REQUEST = 'app/auth/LOGIN_REQUEST';
export const LOGIN_SUCCESS = 'app/auth/LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'app/auth/LOGIN_FAILURE';

export const loginRequest = createAction(LOGIN_REQUEST, (email, password) => Map({ email, password }));
export const loginSuccess = createAction(LOGIN_SUCCESS);
export const loginFailure = createAction(LOGIN_FAILURE);
