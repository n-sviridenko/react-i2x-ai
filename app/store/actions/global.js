import { createAction } from 'redux-actions';

export const SET_LOCALE = 'app/global/SET_LOCALE';
export const SET_TOKEN = 'app/global/SET_TOKEN';
export const UNSET_TOKEN = 'app/global/UNSET_TOKEN';

export const setLocale = createAction(SET_LOCALE);
export const setToken = createAction(SET_TOKEN);
export const unsetToken = createAction(UNSET_TOKEN);
