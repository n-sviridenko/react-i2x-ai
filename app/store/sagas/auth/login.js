import { put, call } from 'redux-saga/effects';

import sendRequest from 'store/sagas/api/sendRequest';
import * as actions from 'store/actions/auth';

export default function* login({ payload }) {
  const data = {
    email: payload.get('email'),
    password: payload.get('password'),
  };

  try {
    const res = yield call(sendRequest, 'POST', 'core/login', data);

    yield put(actions.loginSuccess(res));
  } catch (err) {
    yield put(actions.loginFailure(err));
  }
}
