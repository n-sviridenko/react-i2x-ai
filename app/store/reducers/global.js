import { handleActions } from 'redux-actions';
import { fromJS } from 'immutable';
import { createSelector } from 'reselect';

import { defaultLocale } from 'config';
import { SET_LOCALE, SET_TOKEN, UNSET_TOKEN } from 'store/actions/global';

const initialState = fromJS({
  locale: defaultLocale,
  token: null,
});

export default handleActions({
  [SET_LOCALE]: (state, { payload }) => state.set('locale', payload),
  [SET_TOKEN]: (state, { payload }) => state.set('token', payload),
  [UNSET_TOKEN]: (state) => state.set('token', null),
}, initialState);

export const getRoot = (state) => state.get('global');
export const getLocale = createSelector(getRoot, (state) => state.get('locale'));
export const getToken = createSelector(getRoot, (state) => state.get('token'));
