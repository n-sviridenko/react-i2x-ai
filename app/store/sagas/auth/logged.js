import { put, call } from 'redux-saga/effects';

import * as jwt from 'utils/jwt';
import * as actions from 'store/actions/global';

function* logged({ payload }) {
  const token = payload.get('token');
  const time = yield call(Date.now);

  if (yield call(jwt.isValid, token, time)) {
    yield put(actions.setToken(token));
  }
}

export default logged;
