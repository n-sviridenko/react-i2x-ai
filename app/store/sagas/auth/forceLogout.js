import { select, put } from 'redux-saga/effects';

import { getToken } from 'store/reducers/global';
import { unsetToken } from 'store/actions/global';

export default function* forceLogout() {
  const token = yield select(getToken);

  // in case if user was authenticated
  if (token !== null) {
    yield put(unsetToken());
  }
}
