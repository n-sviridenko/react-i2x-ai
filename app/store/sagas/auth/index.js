import { takeLatest } from 'redux-saga/effects';

import * as actions from 'store/actions/auth';
import login from './login';
import logged from './logged';

export function* loginWatcher() {
  yield takeLatest(actions.LOGIN_REQUEST, login);
}

export function* loggedWatcher() {
  yield takeLatest(actions.LOGIN_SUCCESS, logged);
}

export default [
  loginWatcher,
  loggedWatcher,
];
