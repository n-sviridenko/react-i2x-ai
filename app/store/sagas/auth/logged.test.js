/* eslint-disable redux-saga/yield-effects */
import { put } from 'redux-saga/effects';
import { fromJS } from 'immutable';

import * as actions from 'store/actions/auth';
import * as globalActions from 'store/actions/global';
import logged from './logged';

describe('logged', () => {
  let loggedTask;

  beforeEach(() => {
    const res = fromJS({ token: 'yyyy' });
    loggedTask = logged(actions.loginSuccess(res));

    const dateDescriptor = loggedTask.next().value;
    expect(dateDescriptor).toMatchSnapshot();
    const ifDescriptor = loggedTask.next(new Date(2011, 1, 11, 0, 45, 15)).value;
    expect(ifDescriptor).toMatchSnapshot();
  });

  it('should dispatch setToken if the token is valid', () => {
    const putDescriptor = loggedTask.next(true).value;
    expect(putDescriptor).toEqual(put(globalActions.setToken('yyyy')));
  });

  it('should not dispatch setToken if the token is invalid', () => {
    const isDone = loggedTask.next(false).done;
    expect(isDone).toBe(true);
  });
});
