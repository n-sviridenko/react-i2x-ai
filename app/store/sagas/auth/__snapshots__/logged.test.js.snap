exports[`logged should dispatch setToken if the token is valid 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`logged should dispatch setToken if the token is valid 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "yyyy",
      2011-02-10T23:45:15.000Z,
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`logged should not dispatch setToken if the token is invalid 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`logged should not dispatch setToken if the token is invalid 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "yyyy",
      2011-02-10T23:45:15.000Z,
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
